src/memory/AmdFlash.cc

```diff
@@ -211,32 +215,73 @@ void AmdFlash::setState(State newState)
 
 uint8_t AmdFlash::peek(size_t address) const
 {
-	auto [sector, sectorSize, offset] = getSectorInfo(address);
 	if (state == State::IDLE) {
+		auto [sector, sectorSize, offset] = getSectorInfo(address);
 		if (const uint8_t* addr = readAddress[sector]) {
 			return addr[offset];
 		} else {
 			return 0xFF;
 		}
-	} else {
+	} else if (state == State::IDENT) {
 		if (addressing == Addressing::BITS_12) {
+			if (chip.autoSelect.oddZero && (address & 1)) {
+				// some devices return zero for odd bits instead of mirroring
+				return 0x00;
+			}
 			// convert the address to the '11 bit case'
 			address >>= 1;
 		}
-		switch (address & 3) {
-		case 0:
-			return narrow_cast<uint8_t>(ID >> 8);
-		case 1:
-			return narrow_cast<uint8_t>(ID & 0xFF);
-		case 2:
-			// 1 -> write protected
+		return narrow_cast<uint8_t>(peekAutoSelect(address, chip.autoSelect.undefined));
+	} else {
+		UNREACHABLE;
+	}
+}
+
+uint16_t AmdFlash::peekAutoSelect(size_t address, uint16_t undefined) const
+{
+	switch (address & chip.autoSelect.readMask) {
+	case 0x0:
+		return to_underlying(chip.autoSelect.manufacturer);
+	case 0x1:
+		return chip.autoSelect.device[0] | 0x2200;
+	case 0x2:
+		{
+			if (addressing == Addressing::BITS_12) {
+				// convert the address from the '11 bit case'
+				address <<= 1;
+			}
+			auto [sector, sectorSize, offset] = getSectorInfo(address);
 			return isSectorWritable(sector) ? 0 : 1;
-		case 3:
-		default:
-			// TODO what is this? According to this it reads as '1'
-			//  http://www.msx.org/forumtopicl8329.html
-			return 1;
 		}
+	case 0x3:
+		// On AM29F040 it indicates "Autoselect Device Unprotect Code".
+		// Datasheet does not elaborate. Value is 0x01 according to
+		//  https://www.msx.org/forum/semi-msx-talk/emulation/matra-ink-emulated
+		//
+		// On M29W640G it indicates “Extended Block Verify Code”.
```

Comment:
Detail:
I'll only mention this once, but I saw it in several locations, including in commit messages.
Depending on in what editor/terminal I view your patch, these "double quotes" are shown either as
* Regular double quotes (as-if ASCII char 34).
* As question marks or as `<E2><80<9C>`.
* Or as the actual unicode character \u201C left double quotation mark .

I agree that typographically \u201C may be better. But IMHO for comments in source code or for commit message, this does not bring enough advantages over the disadvantages (that not all tools show this correctly). Why not use the regular ASCII double quote character?

So far I've seen this problem for:
* \u2019 right single quotation mark
* \u201C left double quotation mark
* \u201D right double quotation mark

Reply:
Ah, a matter of habit, to me typing these is muscle memory at this point. I’ll change the CPP comments for sure, but hrmrm if I change the commit messages today some’ll sneak back in later I fear…